### DISCUSSION TITLE
How to enforce HSTS

### DISCUSSION THREAD

**Discussion by nichols-green** [directus/directus#11806] [Archive]
I noticed that Directus supports the CONTENT_SECURITY_POLICY portion of Helmet. Does Directus support other sections of Helmet, specifically the HSTS settings? I've tried some guesswork with my config but so far I'm not seeing the strict header when I run the site. Thanks!

**Comment by azrikahar** [comment]
There is no support via environment variables currently, but you can achieve this using [custom API hooks](https://docs.directus.io/extensions/hooks/)!

Here's a brief rundown:

1. [Create a hook extension](https://docs.directus.io/extensions/creating-extensions/), I'll name it as `helmet` in my example
2. Go to the `src/index.js` of your new `helmet` project, and add the following code:
   
    ```js
    import helmet from "helmet";

    export default ({ init }) => {
      init("app.before", ({ app }) => {
        app.use(
          // this is just a simple example
          helmet.hsts({
            maxAge: 123456,
          })
        );
      });
    };
    ```
   
   Notes:  
   - `app` here is the Express app, so you can do what you can usually do with Express in general
   - Reference for init events: https://docs.directus.io/extensions/hooks/#init-events
   - We'll use `app.before` since it happens right after the helmet CSP code: https://github.com/directus/directus/blob/2938821be05eaf195872c34eed709ac9b4a430b4/api/src/app.ts#L95-L121

3. Build the extension to get the `dist/index.js` file. You can copy my example built file if you want to try it out:

    ```js
    // dist/index.js
    
    "use strict";function e(e){return e&&"object"==typeof e&&"default"in e?e:{default:e}}var t=e(require("helmet"));module.exports=({init:e})=>{e("app.before",(({app:e})=>{e.use(t.default.hsts({maxAge:123456}))}))};
    ```
    
4. Place this file in your Directus `extensions/hooks/helmet/index.js` path.
5. You should see `✨ Loaded extensions: helmet` when you start your Directus instance. (or restart if it's already running, or enable [`EXTENSIONS_AUTO_RELOAD`](https://docs.directus.io/configuration/config-options/#extensions))

## Result

### Before

![chrome_0sFlx9gt8l](https://user-images.githubusercontent.com/42867097/155265160-f295e6bb-f378-4ed9-9c53-64334dab322e.png)

### After
  
![chrome_ImURNLIadI](https://user-images.githubusercontent.com/42867097/155265156-7766b1da-2060-4da0-b6e7-bf025ce75b12.png)

**Reply by nichols-green** [reply]
Thanks azrikahar, I'll try the extension route and see if that works for me.

**Reply by nichols-green** [reply]
This worked for me, thanks again!
